File: tex/code/graph_representation.go
type Node struct {
	osmID    int64
	lon, lat float64

	adj map[int]int    // == adjecent edges/nodes
}

type Edge struct {
	osmID       int64
	length      int    // == edge weight
	drivingTime uint
	maxSpeed    uint8
}

type Graph struct {
	nodes []Node
	edges []Edge

	nodeIdx, edgeIdx map[int64]int
}

type DijkstraGraph struct {
	g *Graph
	pq PriorityQueue

	dist    []uint
	parents []int
}

File: tex/code/graph_representation.c
struct node_t
{
    long osm_id;
    double lon, lat;

    GHashTable *adj; // == adjecent edges/nodes
};

struct edge_t
{
    long osm_id;
    int length;
    int max_speed;
    int driving_time; // == edge weight
};

struct graph_t
{
    int n_nodes, n_edges;
    node *nodes;
    edge *edges;

    GHashTable *node_idx;
    GHashTable *edge_idx;
};

struct dgraph_t
{
    graph g;
    pqueue pq;

    int cur; // == index of current node to explore
    int *dist;
    int *parents;
};
